lib/Parse/ParseDecl.cpp

```diff
@@ -5621,8 +5621,56 @@ void Parser::parseAbstractFunctionBody(AbstractFunctionDecl *AFD) {
     Context.Stats->getFrontendCounters().NumFunctionsParsed++;
 
   ParserResult<BraceStmt> Body = parseBraceItemList(diag::invalid_diagnostic);
-  if (!Body.isNull())
-    AFD->setBody(Body.get());
+  if (!Body.isNull()) {
+    // If the body consists of a single expression, turn it into a return 
+    // statement.
+    BraceStmt * BS = Body.get();
+    AFD->setBody(BS);
+    // FIXME: FIXME_NOW: DETERMINE: Do we need to handle the code completion 
+    //        case here?  It is being handled for closure 
+    //        (Parser::parseExprClosure).
+    if (BS->getNumElements() == 1) {
+      auto Element = BS->getElement(0);
+      // FIXME: FIXME_NOW: DETERMINE: Do we need to do this?  It's being done
+      //        for closures (Parser::parseExprClosure) but it may not be 
+      //        relevant here.
+      //
+      //           if (Element.is<Stmt *>()) {
+      //             if (auto returnStmt =
+      //                         dyn_cast_or_null<ReturnStmt>(Element.get<Stmt*>())) {
+      //               
+      //               if (!returnStmt->hasResult()) {
+      //                 Expr *returnExpr = TupleExpr::createEmpty(Context,
+      //                                                           SourceLoc(),
+      //                                                           SourceLoc(),
+      //                                                           /*implicit*/true);
+      //                 
+      //                 returnStmt->setResult(returnExpr);
+      //                 AFD->setHasSingleExpressionBody();
+      //                 AFD->setSingleExpressionBody(returnExpr);
+      //               }
+      //             }
+      //           } else 
+      if (Element.is<Expr *>()) {
+        auto E = Element.get<Expr *>();
+        if (auto F = dyn_cast_or_null<FuncDecl>(AFD)) {
```

Comment:
Nitpick: The `or_null` variant is for when the value might be null, not what you want as output. `dyn_cast` is fine here.

Reply:
Ah ha!  These usages of `dyn_cast_or_null` are gone in 
https://github.com/apple/swift/pull/23251/commits/a310279dabda30f1de50b2a6bb2c11c0ef20868f .